api/capi/src/nnstreamer-capi-single.c

```diff
@@ -352,6 +356,7 @@ ml_single_close (ml_single_h single)
   }
 
   single_h = (ml_single *) single;
+  g_mutex_lock (&single_h->lock);
```

Comment:
Why do we have to append mutex facility? Is this statement is a critical section to care the consistency of single_h?

Reply:
right, for the variables in handle to prevent accessing the pipeline in a time.

Reply:
Which variables do you target to protect?
In general, it appears that you are defining "critical section" too widely.

Reply:
To prevent invoke() in a time (single pipeline input)

Reply:
Uh.. then, let's add "valid" variable in ```single_h```, set it true at "open" and set it false at "close" before doing unlock(). Otherwise, invoke might enter between Line 381 and 382.

Reply:
Add new flag 'is_valid' and check the valid state